web_ui/prometheus.go

```diff
@@ -138,6 +143,83 @@ func runtimeInfo() (api_v1.RuntimeInfo, error) {
 	return api_v1.RuntimeInfo{}, nil
 }
 
+func checkPromToken(av1 *route.Router) gin.HandlerFunc {
+	/* A function which wraps around the av1 router to force a jwk token check using
+	 * the origin's private key. It will check the request's URL and Header for a token
+	 * and if found it will then attempt to validate the token. If valid, it will continue
+	 * the routing as normal, otherwise it will return an error"
+	 */
+	return func(c *gin.Context) {
+		req := c.Request
+
+		var strToken string
+		var token jwt.Token
+		var err error
+		if authzQuery := req.URL.Query()["authz"]; len(authzQuery) > 0 {
+			strToken = authzQuery[0]
+		} else if authzHeader := req.Header["Authorization"]; len(authzHeader) > 0 {
+			strToken = strings.TrimPrefix(authzHeader[0], "Bearer ")
+		} else {
+			c.JSON(403, gin.H{"error": "Permission Denied: Missing token"})
+			return
+		}
+
+		token, err = jwt.Parse([]byte(strToken), jwt.WithVerify(false))
+
+		if err != nil {
+			c.JSON(400, gin.H{"error": "Failed to parse token"})
+		}
+
+		fedURL := viper.GetString("FederationURL")
+
+		var bKey *jwk.Key
+		if fedURL == token.Issuer() {
+			err := pelican_config.DiscoverFederation()
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to discover the federation information"})
+			}
+			fedURIFile := viper.GetString("FederationURI")
+			response, err := http.Get(fedURIFile)
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to get federation key file"})
+			}
+			defer response.Body.Close()
+			contents, err := io.ReadAll(response.Body)
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to read federation key file"})
+				return
+			}
+			key, err := jwk.ParseKey(contents, jwk.WithPEM(true))
```

Comment:
A few notes about this:

- The `jwk.ParseKey` function will fail if contents has more than one key. While the federation only makes available a single public key at the moment, I don't know if that will always be the case. Can you look at doing this with `jwk.Parse()` so we make sure to always look at the entire keyset?

- I'm not sure this should be passed with the `jwk.WithPEM` option, as the Federation's jwks isn't stored in PEM format. Does it work anyway?

Reply:
I think I got the withPEM from some other parsing we do, probably based on the origin, but I can change this